### PR TITLE
Use a hash map for specs. If the request is a point lookup then just …

### PR DESCRIPTION
**Pull request by baldersheim**
…use a hash lookup.

If it is a wildcard lookup iterate as earlier on.
Also use vespalib::stringref in interface to avoid conversion.
Use vespalib:string in the hash map to locate string in object aswe are still on old abi.

@havardpe and @arnej27959 PR

### CODE REVIEW

**Comment by arnej27959** [comment]
I don't see any gain from not using std::string and std::unordered_map here.
Note that the return value from lookup() still uses std::string so there will need to be conversion anyway.
I'm not sure what old ABI you're talking about, we stopped using reference-counted std::string years ago?

**Comment by baldersheim** [comment]
Yes, we did as long as long as we build our own toolchain. But that stopped on RHEL7/CentOS7.
So we are still using refcounted strings. We must wait for RHEL8 to get usable strings.

**Comment by arnej27959** [comment]
> We must wait for RHEL8 to get usable strings.

NOOOOOO !!!!
bleh.  In that case we should replace all std::string with vespalib::string in all of slobrok,
I will do that in a separate PR.

**Comment by havardpe** [comment]
> Yes, we did as long as long as we build our own toolchain. But that stopped on RHEL7/CentOS7.
> So we are still using refcounted strings. We must wait for RHEL8 to get usable strings.

This would also mean that search performance is suffering (again) due to protobuf using std::string for things like document id...

File: slobrok/src/vespa/slobrok/sbmirror.cpp [hunk 1]
```diff
@@ -167,34 +174,26 @@ MirrorAPI::handleIncrementalFetch()
...
+        updateTo(std::move(specs), diff_to);
     } else if (_specsGen == diff_from) {
         // incremental update
-        SpecList specs;
-        SpecList::const_iterator end = _specs.end();
-        for (SpecList::const_iterator it = _specs.begin();
-             it != end;
-             ++it)
-        {
+        SpecMap specs;
```
Comment (havardpe) [inline slobrok/src/vespa/slobrok/sbmirror.cpp:184]: might be faster to just copy and modify based on diff now that we have a map...

**Pull request merged**